<h1><center> CloudAEye MultiClass Classifier Part-1 </center></h1>

Connecting And Mounting Google Drive

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
%cd /content/drive/My\ Drive/CloudAEye

/content/drive/My Drive/CloudAEye


Importing Libraries and Reading Dataframe


In [39]:
import pandas as pd
import numpy as np

from sklearn import model_selection
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder

import xgboost as xgb
from xgboost import XGBClassifier


In [4]:
df= pd.read_csv('consolidated-metrics.csv')
df.head()

timestamp  ...                       status
0  2020-09-18 21:45:03  ...                       NORMAL
1  2020-09-18 21:46:03  ...  CPU_BURN_KANBAN_API_GATEWAY
2  2020-09-18 21:47:03  ...  CPU_BURN_KANBAN_API_GATEWAY
3  2020-09-18 21:48:03  ...  CPU_BURN_KANBAN_API_GATEWAY
4  2020-09-18 21:49:03  ...  CPU_BURN_KANBAN_API_GATEWAY

[5 rows x 97 columns]

In [5]:
df.shape

(4294, 97)

In [6]:
df['status'].value_counts()

NORMAL                                  2757
NETWORK_DELAY_KANBAN_API_GATEWAY         646
NETWORK_DELAY_KANBAN_COMMAND_SERVICE     500
POD_KILL_KANBAN_API_GATEWAY              310
CPU_BURN_KANBAN_API_GATEWAY               80
Name: status, dtype: int64

Taking a look at Missing Data

In [7]:
percent_missing = df.isnull().sum() * 100 / len(df)
missing_value_df = pd.DataFrame({'column_name': df.columns,
                                 'percent_missing': percent_missing})
missing_value_df.sort_values('percent_missing', inplace=True,ascending=False)

In [15]:
missing_value_df.head(20)

column_name  percent_missing
kubelet_pod_worker_duration_bucket        kubelet_pod_worker_duration_bucket        94.853284
kanban_tasks_get_history                            kanban_tasks_get_history         4.913833
kanban_tasks_complete_with_200_status  kanban_tasks_complete_with_200_status         4.913833
kafka_log_size_growth_rate                        kafka_log_size_growth_rate         0.419190
kubelet_volume_stats_used_bytes              kubelet_volume_stats_used_bytes         0.116442
kubelet_volume_stats_inodes_used            kubelet_volume_stats_inodes_used         0.116442
kubelet_volume_stats_inodes_free            kubelet_volume_stats_inodes_free         0.116442
kubelet_volume_stats_available_bytes    kubelet_volume_stats_available_bytes         0.116442
kafka_network_transmit                                kafka_network_transmit         0.093153
kafka_network_receive                                  kafka_network_receive         0.093153
http_request_duration_microseconds        http_request_duration_microseconds         0.069865
mysql_bytes_received                                    mysql_bytes_received         0.069865
http_response_size_bytes                            http_response_size_bytes         0.069865
mem_free                                                            mem_free         0.046577
mongodb_network_in                                        mongodb_network_in         0.046577
httpsessions_active                                      httpsessions_active         0.046577
kanban_boards_creation                                kanban_boards_creation         0.046577
aws_network_out                                              aws_network_out         0.046577
aws_ec2_network_out_average                      aws_ec2_network_out_average         0.046577
aws_ec2_cpuutilization_average                aws_ec2_cpuutilization_average         0.046577

In [9]:
tempdf1 = df[df['mysql_bytes_received'].isna()]
tempdf1['status'].unique()

array(['NORMAL', nan], dtype=object)

There is only one column named 'kubelet_pod_worker_duration_bucket' which has more than 94%missing data. Therefore removing this column.

Dropping column with high missing data

In [10]:
columns_to_remove  = list(missing_value_df[missing_value_df.percent_missing >50].column_name)
df.drop(columns_to_remove,axis=1,inplace=True)

In [11]:
df.shape

(4294, 96)

Also, we have some missing values in Target variable. Therfore removing those

In [18]:
df = df[df['status'].notna()]

In [19]:
df.shape

(4293, 96)

In [20]:
df['status'].unique()

array(['NORMAL', 'CPU_BURN_KANBAN_API_GATEWAY',
       'NETWORK_DELAY_KANBAN_API_GATEWAY',
       'NETWORK_DELAY_KANBAN_COMMAND_SERVICE',
       'POD_KILL_KANBAN_API_GATEWAY'], dtype=object)

Taking a look at columns that are not integer or float type

In [22]:
for i in df.columns:
  if df[i].dtype != np.number:
    print(i)

timestamp
status


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: DeprecationWarning: Converting `np.inexact` or `np.floating` to a dtype is deprecated. The current result is `float64` which is not strictly correct.
  


# I will be using XGboost here due to reasons as below
## Since there is Missing data in few column. Xgboost will be able to handle it internally.
## No need to scale or transform the data.
## Since this is an ensemble model.Correlation will not be an issue here 

## Preparing Data ForXGBOOST

In [28]:
df_xg = df.iloc[:,1:].copy()
df_xg.head()

apiserver_current_inflight_requests  ...                       status
0                             0.008333  ...                       NORMAL
1                             0.008333  ...  CPU_BURN_KANBAN_API_GATEWAY
2                             0.008333  ...  CPU_BURN_KANBAN_API_GATEWAY
3                             0.016138  ...  CPU_BURN_KANBAN_API_GATEWAY
4                             0.020305  ...  CPU_BURN_KANBAN_API_GATEWAY

[5 rows x 95 columns]

In [30]:
missing_indicator = -1000
df_xg.fillna(missing_indicator, inplace=True)

Using label encoder for the Target variable 'Status'

In [33]:
# encode string class values as integers


label_encoder = LabelEncoder()
label_encoder = label_encoder.fit(df_xg['status'])
label_encoded_y = label_encoder.transform(df_xg['status'])

In [34]:
df_xg.drop('status',axis=1,inplace=True)

In [36]:
seed = 7
test_size = 0.20
X_train, X_test, y_train, y_test = model_selection.train_test_split(df_xg, label_encoded_y, test_size=test_size, random_state=seed)

Creating XGboost Dmatrix

In [38]:
# convert data to xgb data structure
xgtrain = xgb.DMatrix(X_train, y_train, 
                        missing=missing_indicator)
xgtest = xgb.DMatrix(X_test, label=y_test, 
                        missing=missing_indicator)

In [43]:
xgb_num_rounds = 720
def get_params():
    
    params = {}
    params["objective"] = "reg:linear"     
    params["eta"] = 0.05
    params["min_child_weight"] = 360
    params["subsample"] = 0.85
    params["colsample_bytree"] = 0.3
    params["silent"] = 1
    params["max_depth"] = 7
    plst = list(params.items())

    return plst

In [47]:
def eval_wrapper(yhat, y):  
    y = np.array(y)
    y = y.astype(int)
    yhat = np.array(yhat)
    yhat = np.clip(np.round(yhat), np.min(y), np.max(y)).astype(int)   
    accuracy_score(y, yhat)

In [42]:
# get the parameters for xgboost
plst = get_params()
print(plst) 

[('objective', 'reg:linear'), ('eta', 0.05), ('min_child_weight', 360), ('subsample', 0.85), ('colsample_bytree', 0.3), ('silent', 1), ('max_depth', 7)]


Training the model

In [44]:
# train model
model = xgb.train(plst, xgtrain, xgb_num_rounds) 

In [62]:
# get preds
train_preds = model.predict(xgtrain, ntree_limit=model.best_iteration)
tp_round = train_preds.round().astype(int)
print('Train score is:', eval_wrapper(tp_round, y_train)) 
test_preds = model.predict(xgtest, ntree_limit=model.best_iteration)

Train score is: None


In [64]:
test_prediction = test_preds.round().astype(int)

Predicting Scores

In [63]:
accuracy_score(y_train, tp_round)

0.8948747815958067

In [65]:
accuracy_score(y_test, test_prediction)


0.8323632130384168

## Since i dont have enough infromation about this data and its importance. I used accuracy as the metric for judgement.
Based on the results the test accuracy is over **83%** and train accuracy is over **89%**